### ISSUE TITLE
[SMASH] "parse error: Invalid numeric literal" instead of "Pool XYZ is delisted" when querying delisted pool

### ISSUE THREAD

**Issue by ArturWieczorek** [IntersectMBO/cardano-db-sync#1132]
**OS**
Your OS:
Linux Mint

**Versions**
The `db-sync` version (eg `cardano-db-sync --version`):

tag: 13.0.0-rc1
commit c59164c88b6f7e1d4d389c3f66de2d3157e3c29e 

**Build/Install Method**
`nix`

**Run method**

> PGPASSFILE=config/pgpass-testnet smash-server/bin/cardano-smash-server --config /nix/store/5b6pry15w93fv0r0x9rc3r1ii5871lvr-config-0-0.json --port 3100 --admins admins.txt

**Additional context**
N/A

**Problem Report**
Querying delisted pool (`smash v13`) returns:

`parse error: Invalid numeric literal` 

instead of: 

`Pool XYZ is delisted` as it was in `smash v12`.

**Steps to reproduce**

This was performed on `testnet`.

a) Query some pool:

```sh
curl -X GET -v http://localhost:3100/api/v1/metadata/81e84003f3d2f65315b479dc3cdbe4aa8c8595a3d76818e284b29f27/8e123d238a44a91bab56ebd631020bede5a381d8487a46d1f2ebbdc52e3b3eba | jq .

{
  "name": "OZZIE",
  "ticker": "OZZIE",
  "homepage": "https://ozzieoffsec.github.io/pool",
  "description": "We love Turtles"
}
```

b) Delist it:

```sh
curl --verbose -u username:password --header "Content-Type: application/json" --request PATCH --data '{"poolId":"81e84003f3d2f65315b479dc3cdbe4aa8c8595a3d76818e284b29f27"}' http://localhost:3100/api/v1/delist

{"poolId":"81e84003f3d2f65315b479dc3cdbe4aa8c8595a3d76818e284b29f27"}
```
```sql
testnet=# select * from delisted_pool;
 id |                          hash_raw
----+------------------------------------------------------------
  1 | \x81e84003f3d2f65315b479dc3cdbe4aa8c8595a3d76818e284b29f27
(1 row)
```

c) Query pool after it was delisted:

Response from curl:

1) `smash v12`:
```sh
curl -X GET -v http://localhost:3100/api/v1/metadata/81e84003f3d2f65315b479dc3cdbe4aa8c8595a3d76818e284b29f27/8e123d238a44a91bab56ebd631020bede5a381d8487a46d1f2ebbdc52e3b3eba | jq .

Pool 81e84003f3d2f65315b479dc3cdbe4aa8c8595a3d76818e284b29f27 is delisted
```

2) `smash v13`:

```
parse error: Invalid numeric literal at line 1, column 5
```

Response from `smash-server` (for both `smash v12` &`smash v13`):
```
[smash-server:Warning:40] [2021-11-25 10:02:36.40 UTC] Pool 81e84003f3d2f65315b479dc3cdbe4aa8c8595a3d76818e284b29f27 is delisted
```

**Comment by kderme** [comment]
Seems like a json parser error. We had to change aeson revision recently to match the used revision upstream.

**Comment by tmcgilchrist** [comment]
@ArturWieczorek I've tried to reproduce this bug using docker containers and haven't managed to. Here is what I've tried

## With cardano-db-sync:12.0.2 and cardano-node:1.34.1 in docker-compose.yml and SMASH from tag:12.0.2
``` shell
RESTORE_SNAPSHOT=https://update-cardano-mainnet.iohk.io/cardano-db-sync/12/db-sync-snapshot-schema-12-block-7107980-x86_64.tgz \
NETWORK=mainnet docker-compose up && docker-compose logs -f
```

Select a suitable pool to delist:

``` shell
$ select * from pool_hash join pool_metadata_ref on pool_hash.id = pool_metadata_ref.id;
  id  |                          hash_raw                          |                           view                           |  id  | pool_id |                               url                                |                                hash                                | registered_tx_id 
------+------------------------------------------------------------+----------------------------------------------------------+------+---------+------------------------------------------------------------------+--------------------------------------------------------------------+------------------
    2 | \x0f292fcaa02b8b2f9b3c8f9fd8e0bb21abedb692a6d5058df3ef2735 | pool1pu5jlj4q9w9jlxeu370a3c9myx47md5j5m2str0naunn2q3lkdy |    2 |       2 | https://stakenuts.com/mainnet.json                               | \x47c0c68cb57f4a5b4a87bad896fc274678e7aea98e200fa14a1cb40c0cab1d8c |          2415602

```
Request the Metadata 

``` shell
$ curl --verbose --header "Content-Type: application/json" \
--request GET http://localhost:3100/api/v1/metadata/0f292fcaa02b8b2f9b3c8f9fd8e0bb21abedb692a6d5058df3ef2735/47c0c68cb57f4a5b4a87bad896fc274678e7aea98e200fa14a1cb40c0cab1d8c | jq .
Note: Unnecessary use of -X or --request, GET is already inferred.
*   Trying 127.0.0.1:3100...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to localhost (127.0.0.1) port 3100 (#0)
> GET /api/v1/metadata/0f292fcaa02b8b2f9b3c8f9fd8e0bb21abedb692a6d5058df3ef2735/47c0c68cb57f4a5b4a87bad896fc274678e7aea98e200fa14a1cb40c0cab1d8c HTTP/1.1
> Host: localhost:3100
> User-Agent: curl/7.79.1
> Accept: */*
> Content-Type: application/json
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Transfer-Encoding: chunked
< Date: Fri, 24 Jun 2022 06:17:17 GMT
< Server: Warp/3.3.17
< Content-Type: application/json;charset=utf-8
< Cache-Control: no-store
< 
{ [184 bytes data]
100   171    0   171    0     0   2012      0 --:--:-- --:--:-- --:--:--  2192
* Connection #0 to host localhost left intact
{
  "name": "StakeNuts",
  "description": "StakeNuts.com",
  "ticker": "NUTS",
  "homepage": "https://stakenuts.com/",
  "extended": "https://stakenuts.com/mainnet-extended.json"
}

```

Delist that pool

``` shell
$ curl --verbose -u tmcgilchrist:tmcgilchrist --header "Content-Type: application/json" --request PATCH --data '{"poolId": "0f292fcaa02b8b2f9b3c8f9fd8e0bb21abedb692a6d5058df3ef2735"}' http://localhost:3100/api/v1/delist
*   Trying 127.0.0.1:3100...
* Connected to localhost (127.0.0.1) port 3100 (#0)
* Server auth using Basic with user 'tmcgilchrist'
> PATCH /api/v1/delist HTTP/1.1
> Host: localhost:3100
> Authorization: Basic dG1jZ2lsY2hyaXN0OnRtY2dpbGNocmlzdA==
> User-Agent: curl/7.79.1
> Accept: */*
> Content-Type: application/json
> Content-Length: 70
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Transfer-Encoding: chunked
< Date: Fri, 24 Jun 2022 06:19:07 GMT
< Server: Warp/3.3.17
< Content-Type: application/json;charset=utf-8
< 
* Connection #0 to host localhost left intact
{"poolId":"0f292fcaa02b8b2f9b3c8f9fd8e0bb21abedb692a6d5058df3ef2735"}%    
```

Request Metata again

``` shell
$ curl --verbose --header "Content-Type: application/json" \
--request GET http://localhost:3100/api/v1/metadata/0f292fcaa02b8b2f9b3c8f9fd8e0bb21abedb692a6d5058df3ef2735/47c0c68cb57f4a5b4a87bad896fc274678e7aea98e200fa14a1cb40c0cab1d8c       
Note: Unnecessary use of -X or --request, GET is already inferred.
*   Trying 127.0.0.1:3100...
* Connected to localhost (127.0.0.1) port 3100 (#0)
> GET /api/v1/metadata/0f292fcaa02b8b2f9b3c8f9fd8e0bb21abedb692a6d5058df3ef2735/47c0c68cb57f4a5b4a87bad896fc274678e7aea98e200fa14a1cb40c0cab1d8c HTTP/1.1
> Host: localhost:3100
> User-Agent: curl/7.79.1
> Accept: */*
> Content-Type: application/json
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 403 Forbidden
< Transfer-Encoding: chunked
< Date: Fri, 24 Jun 2022 06:19:48 GMT
< Server: Warp/3.3.17
< 
* Connection #0 to host localhost left intact
Pool 0f292fcaa02b8b2f9b3c8f9fd8e0bb21abedb692a6d5058df3ef2735 is delisted%       
```

Note the response has a content type of `application/json` but the content is a plain text string and will cause a `jq` parse error. 

## With cardano-db-sync:13.0.0-rc1 and cardano-node:1.34.1 in docker-compose.yml and SMASH from tag:13.0.0-rc1
 
``` shell
RESTORE_SNAPSHOT=https://update-cardano-mainnet.iohk.io/cardano-db-sync/13/db-sync-snapshot-schema-13-block-6429999-x86_64.tgz \
NETWORK=mainnet docker-compose up && docker-compose logs -f
```

Find a suitable pool to delist:
```
$ select * from pool_hash join pool_metadata_ref on pool_hash.id = pool_metadata_ref.id;
  id  |                          hash_raw                          |                           view                           |  id  | pool_id |                               url                                |                                hash                                | registered_tx_id 
------+------------------------------------------------------------+----------------------------------------------------------+------+---------+------------------------------------------------------------------+--------------------------------------------------------------------+------------------
    1 | \x153806dbcd134ddee69a8c5204e38ac80448f62342f8c23cfe4b7edf | pool1z5uqdk7dzdxaae5633fqfcu2eqzy3a3rgtuvy087fdld7yws0xt |    1 |       1 | https://raw.githubusercontent.com/Octalus/cardano/master/p.json  | \xca7d12decf886e31f5226b5946c62edc81a7e40af95ce7cd6465122e309d5626 |          2415556
```

Request the Metadata

``` shell
curl --verbose --header "Content-Type: application/json" \ 
  --request GET http://localhost:3100/api/v1/metadata/153806dbcd134ddee69a8c5204e38ac80448f62342f8c23cfe4b7edf/ca7d12decf886e31f5226b5946c62edc81a7e40af95ce7cd6465122e309d5626 | jq .

{
  "name": "OctasPool",
  "description": "Octa's Performance Pool",
  "ticker": "OCTAS",
  "homepage": "https://octaluso.dyndns.org",
  "extended": "https://raw.githubusercontent.com/Octalus/cardano/master/pe.json"
}
```

Delist that pool

``` shell
$ curl --verbose -u tmcgilchrist:tmcgilchrist --header "Content-Type: application/json" --request PATCH --data '{"poolId": "153806dbcd134ddee69a8c5204e38ac80448f62342f8c23cfe4b7edf"}' http://localhost:3100/api/v1/delist                     
*   Trying 127.0.0.1:3100...
* Connected to localhost (127.0.0.1) port 3100 (#0)
* Server auth using Basic with user 'tmcgilchrist'
> PATCH /api/v1/delist HTTP/1.1
> Host: localhost:3100
> Authorization: Basic dG1jZ2lsY2hyaXN0OnRtY2dpbGNocmlzdA==
> User-Agent: curl/7.79.1
> Accept: */*
> Content-Type: application/json
> Content-Length: 70
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Transfer-Encoding: chunked
< Date: Fri, 24 Jun 2022 08:16:01 GMT
< Server: Warp/3.3.17
< Content-Type: application/json;charset=utf-8
< 
* Connection #0 to host localhost left intact
{"poolId":"153806dbcd134ddee69a8c5204e38ac80448f62342f8c23cfe4b7edf"}%         
```

Request Metata again

``` shell
$ curl --verbose --header "Content-Type: application/json"   --request GET http://localhost:3100/api/v1/metadata/153806dbcd134ddee69a8c5204e38ac80448f62342f8c23cfe4b7edf/ca7d12decf886e31f5226b5946c62edc81a7e40af95ce7cd6465122e309d5626      
Note: Unnecessary use of -X or --request, GET is already inferred.
*   Trying 127.0.0.1:3100...
* Connected to localhost (127.0.0.1) port 3100 (#0)
> GET /api/v1/metadata/153806dbcd134ddee69a8c5204e38ac80448f62342f8c23cfe4b7edf/ca7d12decf886e31f5226b5946c62edc81a7e40af95ce7cd6465122e309d5626 HTTP/1.1
> Host: localhost:3100
> User-Agent: curl/7.79.1
> Accept: */*
> Content-Type: application/json
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 403 Forbidden
< Transfer-Encoding: chunked
< Date: Fri, 24 Jun 2022 08:16:40 GMT
< Server: Warp/3.3.17
< 
* Connection #0 to host localhost left intact
Pool 153806dbcd134ddee69a8c5204e38ac80448f62342f8c23cfe4b7edf is delisted%
```

Note the response again has a content-type of `application/json` but returns a plain text response that will also cause a `jq` parse error. 


If you're able to reproduce the smash v12 issue, could you please attach the docker images used or the specific git shas used. Many thanks.

**Comment by ArturWieczorek** [comment]
Thanks for info @tmcgilchrist. I will get back to this once I am done with testing **RC4** for `db-sync` (should be this week).

**Comment by ArturWieczorek** [comment]
@tmcgilchrist Sorry, my bad - it looks like this issue was also present in `v12`, so it is not a new problem. I did **not** pipe the query result to `jq` for `v12` and that's why I could see response (but it was in plain text) and I assumed it's OK.

**Comment by tmcgilchrist** [comment]
@ArturWieczorek I went back and had a look at the previous version of SMASH (1.6.1) before it got merged into `cardano-db-sync` repo. The code returns a 403 when the pool is delisted https://github.com/input-output-hk/smash/blob/1.6.1/smash/src/Cardano/SMASH/Lib.hs#L244-L249
It doesn't seem to have returned a valid JSON response in any of the old versions of SMASH.

As a resolution to this issue I would suggest the behaviour is left as is, none of the other error responses have their messages returned as JSON objects and this case is consistent with how SMASH is returning other errors as `Content-Type: application/json` even if that is incorrect. cc @erikd are you happy with this resolution?
